plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
-      session_mgr.remove( this->shared_from_this(), active_entry );
+            //and then send the block
+            if(block_to_send) {
+               const fc::unsigned_int get_blocks_result_variant_index = block_to_send->is_v1_request ?
+                                                                        state_result(get_blocks_result_v1()).index() :
+                                                                        state_result(get_blocks_result_v0()).index();
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(get_blocks_result_variant_index)));
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(block_to_send->blocks_result_base)));
+
+               //accessing the _logs here violates the rule that those should only be accessed on the main thread. However, we're
+               // only calling get_unpacked_entry() on it which assumes the mutex is held by the locked_decompress_stream. So this is
+               // "safe" in some aspects but can deadlock
+               co_await write_log_entry(block_to_send->trace_stream, trace_log, block_to_send->this_block_num);
+               co_await write_log_entry(block_to_send->state_stream, chain_state_log, block_to_send->this_block_num);
+               if(block_to_send->is_v1_request)
+                  co_await write_log_entry(block_to_send->finality_stream, finality_data_log, block_to_send->this_block_num);
+
+               co_await stream.async_write_some(true, boost::asio::const_buffer());
+            }
+         }
+      });
    }
+
+private:
+   ///these items must only ever be touched by the session's strand
+   Executor                          strand;
+   coro_throwing_stream              stream;
+   coro_nonthrowing_steadytimer      wake_timer;
+   unsigned                          coros_running = 0;
+
+   ///these items must only ever be touched on the main thread
+   std::deque<bool>                  queued_status_requests;  //false for v0, true for v1
+
+   get_blocks_request_v0             current_blocks_request;
+   std::optional<bool>               current_blocks_request_v1_finality; //unset: current request is v0; set means v1; true/false is if finality requested
+   //current_blocks_request is modified with the current state; bind some more descriptive names to items frequently used
+   uint32_t&                         send_credits = current_blocks_request.max_messages_in_flight;
+   chain::block_num_type&            next_block_cursor = current_blocks_request.start_block_num;
+
+   chain::controller&                controller;
+   std::optional<state_history_log>& trace_log;
+   std::optional<state_history_log>& chain_state_log;
+   std::optional<state_history_log>& finality_data_log;
+
+   GetBlockID                        get_block_id;
+   GetBlock                          get_block;
+
+   ///these items might be used on either the strand or main thread
```

Comment:
I appreciate these comments and I don't expect SHIP to be dynamic in the near term but, I have concerns that over time, this comment (and the one above about the strand-only variables) will end up being a thin protection. Perhaps in a future iteration we can figure out a more concreate way of throwing errors quickly if these members are accessed from the wrong context.

Reply:
This is certainly the other edge of the implicit-locking-by-using-strands sword. Definitely be curious to explore some solutions here going forward.

One option in mind it to wrap all these inside a container that validates `running_in_this_thread()` (that exists on both a basic_executor and strand). For example (I'm just pseudo coding here) we'd have something like
```c++
executor_validator<std::deque<bool>>      queued_status_requests(app.executor());
executor_validator<get_blocks_request_v0> current_blocks_request(app.executor());
/// etc
```
and then something like,
```c++
template<typename Executor, typename T>
struct executor_validator {
   executor_validator(Executor check_executor) : check_executor(check_executor) { }

   T& operator*() {
      assert(check_executor.running_in_this_thread());
      return t;
   }

private:
   Executor check_executor;
   T t;
};
```
so something like
```c++
*current_blocks_request = ...
```
would assert if not on running on main thread.

I suspect there will be quite a few traps here in practice.